### ISSUE TITLE
iphone 8: safeAreaInsets is undefined if NavigationContainer has children other than createStackNavigator()

### ISSUE THREAD

**Issue by rumax** [react-navigation/react-navigation#8953]
**Current Behavior**

- focusedOptions.safeAreaInsets is `undefined` and as a result top is `0` (in `@react-navigation/stack/src/views/Stack/CardStack.tsx`)

```
const {
      top = insets.top,
      right = insets.right,
      bottom = insets.bottom,
      left = insets.left,
    } = focusedOptions.safeAreaInsets || {};
```

- Screenshot:

<img width="409" alt="Screenshot 2020-10-15 at 12 49 08" src="https://user-images.githubusercontent.com/2707180/96115364-1c4d6300-0ee7-11eb-9414-327ebf6f7c1c.png">

**Expected Behavior**

- Toolbar is properly rendered and `top` has a proper value (20).

<img width="402" alt="Screenshot 2020-10-15 at 13 05 58" src="https://user-images.githubusercontent.com/2707180/96115805-d1801b00-0ee7-11eb-8dbc-f8adfc42f315.png">

**How to reproduce**

```
import { SafeAreaProvider } from 'react-native-safe-area-context';
import { NavigationContainer } from '@react-navigation/native';

export default function App() {
  return (
    <SafeAreaProvider>
      <NavigationContainer>
        <MainStack/>
      </NavigationContainer>
    </SafeAreaProvider>
  );
}
```

where `MainStack` is

```
const AppStackNavigator = createStackNavigator();
```

with few screens.

<hr>

Meantime if `const AppStackNavigator = createStackNavigator();` and screens defined are rendered directly as a child of `<NavigationContainer>`, toolbar is rendered properly

Toolbar also works properly after reload of the app. So it is the issue only for a fresh start.

**Your Environment**

| software                       | version |
| ------------------------------ | ------- |
| iOS                 | Iphone 8
| @react-navigation/native       | ^5.7.6
| @react-navigation/stack        | ^5.9.3
| react-native-gesture-handler   | ^1.8.0
| react-native-safe-area-context | ^3.1.8
| react-native-screens           | ^2.11.0
| react-native    | 0.60.5

**Comment by rumax** [comment]
Hi @github-actions bot. Asking for a minimal repo is useful but usually not possible, because issues happens in real projects not demos. And unfortunately I cannot share that project.

But I've left en entry point for the bug and where to look. It is line `427` in CardSStack https://github.com/react-navigation/react-navigation/blob/main/packages/stack/src/views/Stack/CardStack.tsx#L427:

```
const {
      top = insets.top,
      right = insets.right,
      bottom = insets.bottom,
      left = insets.left,
    } = focusedOptions.safeAreaInsets || {};
```

You are doing  ` || {};` so you should know a case when it is possible. And I afraid there can be more default values. 

So basically the question is:

- When it is possible that `focusedOptions.safeAreaInsets` is undefined? <<-- I think knowing this information that can highlight the problem and knowing that it will be possible to create simple/minimal repo or understand the problem and fix it. 
- Why default values is just an empty object? It could be:

```
const {
      top = insets.top,
      right = insets.right,
      bottom = insets.bottom,
      left = insets.left,
    } = {
    ...defaultSafeAreaInsets,
    ...focusedOptions.safeAreaInsets,
  };
```

**Comment by satya164** [comment]
> because issues happens in real projects not demos. And unfortunately I cannot share that project or create a minimal project

There's no technical difference between a "demo" and a "real project". I don't see why it wouldn't happen in a demo project.

> When it is possible that focusedOptions.safeAreaInsets is undefined

Always, unless you actually pass `safeAreaInsets` in `options`

> Why default values is just an empty object? It could be:

I don't see how that would make any difference in what the end result will be.

> where defaultSafeAreaInsets represents real, device specific values

Those values are already specified as default values for the destructured variables, there's no reason to duplicate them, duplicating them wouldn't change the output

> The simplest advise for you is to make this focusedOptions.safeAreaInsets a required property, and see where the issue is coming from

It's not related

**Comment by rumax** [comment]
> I don't see why it wouldn't happen in a demo project.

@satya164 are you able to create such a demo project that will cover all cases? As a "member" of the project it should be easy for you, isn't it? Moreover you know the code, I don't. So you know what exactly to do and how to reproduce it in "demo" project.

But I was referring to the bug in your lib and a possible way of fixing it. You pass that `safeAreaInsets ` in many places and at the end you do `|| {}` which means there is no required props check or types check. That's also how you could improve the project.

> Always, unless you actually pass safeAreaInsets in options

That's not true, I never pass safeAreaInsets in options, or your demo project doesn't and some how it works

> I don't see how that would make any difference in what the end result will be.

That is sad

> It's not related
> I don't see how that would make any difference in what the end result will be

That's also why I cannot create a "demo" project for you with that exact bug. Because I do not know the project so good as you, moreover creating such a project can take much more effort at the end than fixing a bug. But I believe that understanding of the project is a key to solving the issue.

**Comment by satya164** [comment]
> are you able to create such a demo project that will cover all cases? As a "member" of the project it should be easy for you, isn't it? Moreover you know the code, I don't. So you know what exactly to do and how to reproduce it in "demo" project.

You are reporting a bug, you should know when the bug occurs. No it's not easy for me because I don't know in what situation the bug occurs, you do. If I knew exactly what to do then I wouldn't be asking you for a repro.

You don't need to know the library to create a demo project, you know when the bug occurs in your app so replicate it. That's the whole point of providing a repro.

> But I was referring to the bug in your lib and a possible way of fixing it

It's literally the options passed to the focused screens and it's supposed to be initialized to a default if user hasn't overridden one. "somehow it works" is because it's supposed to work by initializing default values.

You're saying you don't understand the project, yet when I tell you that it's expected to be undefined most of the time, you disagree with me.

I'm not asking you to debug the library and fix it, I'm asking you to provide a way to reproduce the bug so that "I can debug it".

Regardless, it's looks like an issue with your usage of `SafeAreaProvider` and nothing to do with if `NavigationContainer` has children or not. https://github.com/th3rdwave/react-native-safe-area-context#optimization

I'm closing the issue since it doesn't look like this is a bug in the library. If you believe there's a bug and have a way to reproduce the issue, then open a new issue.

**Comment by rumax** [comment]
> I'm closing the issue since it doesn't look like this is a bug in the library. If you believe there's a bug and have a way to reproduce the issue, then open a new issue.

@satya164 thanks for closing the issue and no hint why it is the issue only for iPhone 8. **That's not a project issue, because the project does not contain any platform specific code to render the screen and toolbar, it works fine with Android and iPhoneX.**

I will also leave that screenshot for you:

<img width="1099" alt="Screenshot 2020-11-29 at 16 32 38" src="https://user-images.githubusercontent.com/2707180/100546272-c21f1c00-3260-11eb-9a0d-7c60945d9c31.png">


There is interesting comment, which is part of the lib, I didn't add it manually. You can also see the dependencies used. I do not say it is related, but if you already have or had (I do not see that comment in sources, only in the lib), there is quire a bug chance that I've faced another buggy "buggy on iOS" behaviour. 

But the issue that I faced was somewhere in

```
const {
      top = insets.top,
      right = insets.right,
      bottom = insets.bottom,
      left = insets.left,
    } = focusedOptions.safeAreaInsets || {};
```

and was related to the missing top value. I expected that you, as an owner of the code, could at least give an advise, why it can be empty.

**Comment by satya164** [comment]
> thanks for closing the issue and no hint
> I expected that you, as an owner of the code, could at least give an advise, why it can be empty

Did you even read the link I shared in my comment? I directly linked you to a solution.

The `top` is `0` because you **need to specify** initial values as mentioned here https://github.com/th3rdwave/react-native-safe-area-context#optimization

Or remove the unnecessary `SafeAreaProvider` that we don't ask you to add in React Navigation docs and let React Navigation handle initial values for safe area.

Regardless, I had to read your code several times to understand what could be happening because you don't even want to provide a repro.

Next time please provide a repro when opening an issue. I understand you're trying to be helpful by trying to find the issue, but posting parts of code and insisting that it has a bug when I tell you that it works as expected isn't helpful. If you had provided a repro, I would've found out the issue with your code in minutes and it would be way more helpful and faster.

> There is interesting comment, which is part of the lib, I didn't add it manually

That's regarding react-native-screens doesn't have anything to do with safe area insets.

**Comment by rumax** [comment]
The link is for `so when rotating the device`, and in the project there is no rotating, layout is fixed. But maybe indeed `there is no delay from the asynchronous bridge`, can be the issue. If I find something useful, that can help with the issue, I'll post it. At this moment I cannot create a demo project with such case (but of course with simple screens and less code), because when I try to, it works.
